imi/src/main/java/org/motechproject/nms/imi/service/impl/CsrValidatorServiceImpl.java

```diff
@@ -0,0 +1,181 @@
+    private SubscriptionService subscriptionService;
+    private LanguageLocationService languageLocationService;
+    private CircleService circleService;
+
+
+    @Autowired
+    public CsrValidatorServiceImpl(@Qualifier("imiSettings") SettingsFacade settingsFacade,
+                                   SubscriptionService subscriptionService,
+                                   LanguageLocationService languageLocationService,
+                                   CircleService circleService) {
+        this.settingsFacade = settingsFacade;
+        this.subscriptionService = subscriptionService;
+        this.languageLocationService = languageLocationService;
+        this.circleService = circleService;
+    }
+
+
+    private void loadPacks() {
+        weekIds = new HashMap<>();
+        contentFileNames = new HashMap<>();
+        for (SubscriptionPack pack : subscriptionService.getSubscriptionPacks()) {
+            Set<String> packWeekIds = new HashSet<String>();
+            Set<String> packContentFileNames = new HashSet<String>();
+            for (SubscriptionPackMessage message :pack.getMessages()) {
+                packWeekIds.add(message.getWeekId());
+                packContentFileNames.add(message.getMessageFileName());
+            }
+            weekIds.put(pack.getName(), packWeekIds);
+            contentFileNames.put(pack.getName(), packContentFileNames);
+        }
+    }
+
+
+    private void loadLanguageLocationCodes() {
+        languageLocationCodes = new HashSet<String>();
+        for (LanguageLocation languageLocation : languageLocationService.getAll()) {
+            languageLocationCodes.add(languageLocation.getCode());
+        }
+    }
+
+
+    private void loadCircles() {
+        circles = new HashSet<String>();
+        for (Circle circle : circleService.getAll()) {
+            circles.add(circle.getName());
+        }
+    }
+
+
+    private void validateWeekId(String pack, String week) {
+        if (weekIds == null) { loadPacks(); }
+        if (weekIds.containsKey(pack)) {
+            if (weekIds.get(pack).contains(week)) {
+                return;
+            }
+            throw new InvalidCsrException(String.format("Invalid weekId for %s pack: %s", pack, week));
+        }
+        throw new IllegalStateException(String.format("Unexpected pack name: %s", pack));
+    }
+
+
+    private void validateContentFileName(String pack, String fileName) {
+        if (contentFileNames == null) { loadPacks(); }
+        if (contentFileNames.containsKey(pack)) {
+            if (contentFileNames.get(pack).contains(fileName)) {
+                return;
+            }
+            throw new InvalidCsrException(String.format("Invalid contentFileName for %s pack: %s", pack, fileName));
+        }
+        throw new IllegalStateException(String.format("Unexpected pack name: %s", pack));
+    }
+
+
+    private void validateCircle(String name) {
+        if (circles == null) { loadCircles(); }
```

Comment:
Good point, I imagine that any update of one of these basic tables should require a MOTECH restart.
But maybe we can have a 'refresh cache' event?